Bluetooth: controller: LLCP: Rejection logic for handling non-supported features

thoh-ot opened the pull request:
Implement rejection logic for handling non-supported features and unit tests for:

- Invalid opcodes based on controller role
- No LE Encryption support
- No Connection Parameters Request procedure support
- No Peripheral-initiated Features Exchange support
- No LE 2M PHY/LE Coded PHY support

Update ztest for non-kernel test to support skipping via `ztest_test_skip()`.

thoh-ot committed:
ztest: Add skipping for non-kernel tests

Add ztest_test_skip() support to non-kernel tests by implementing
long jump buffer and TC_SKIP result collection.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

erbr-ot reviewed (commented):
Looks real good to me. Look forward to trying it and perhaps to get a walk through on Thursday?

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:
```diff
@@ -383,6 +392,25 @@ static void rr_act_reject(struct ll_conn *conn)
+	if (ctx->pause || !llcp_tx_alloc_peek(conn, ctx)) {
+		rr_set_state(conn, RR_STATE_UNSUPPORTED);
+	} else {
+		struct proc_ctx *ctx = llcp_rr_peek(conn);
```
This is a merge error and needs to be fixed

kruithofa commented on tests/bluetooth/controller/ctrl_unsupported/src/main.c:
```diff
@@ -0,0 +1,332 @@
+}
+
+void test_invalid_per_rem(void)
+{
```
add test for undefined opcode

kruithofa replied:
in particular opcode 255 may be an issue

thoh-ot committed:
Bluetooth: controller: Unsupported proc handling

Add handling of unsupported procedures based on opcode and role.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: Test invalid LLCP opcodes

Test the rejection (LL_UNKNOWN_RSP) of invalid remotely initiated
procedures in both the central and peripheral role.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

thoh-ot committed:
tests: Bluetooth: Test unsupported LLCP opcodes

Test the rejection (LL_UNKNOWN_RSP) of unsupported remotely initiated
procedures in both the central and peripheral role.

Unsupported opcodes are handled by Kconfig manipulation and removal of
source files.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

erbr-ot reviewed (dismissed):
LGTM

tejlmand commented on tests/bluetooth/controller/ctrl_unsupported/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
```suggestion
cmake_minimum_required(VERSION 3.20)
```

tejlmand commented on tests/bluetooth/controller/ctrl_unsupported/testcase.yaml:
```diff
@@ -3,3 +3,15 @@ common:
      type: unit
+  bluetooth.controller.ctrl_unsupported.no_enc.test:
+     type: unit
+     extra_args: NO_ENC=y KCONFIG_OVERRIDE_FILE="kconfig_override.h"
```
seems like I was not invited to this party #40073

The way `KCONFIG_OVERRIDE_FILE` and `tests/bluetooth/controller/common/defaults_cmake.txt` appears to be short cutting everything else we do in the build system.

But guess that is outside the scope of this PR.

tejlmand reviewed (dismissed):
CMake version is now 3.20

ppryga-nordic reviewed (approved):
I did quick pass-through and I didn't find anything blocking the PR so approve it.
